letsencrypt/storage.py

```diff
@@ -0,0 +1,427 @@
+
+    def has_pending_deployment(self):
+        """Is there a later version of all of the managed items?"""
+        # TODO: consider whether to assume consistency or treat
+        #       inconsistent/consistent versions differently
+        smallest_current = min(self.current_version(x) for x in ALL_FOUR)
+        return smallest_current < self.latest_common_version()
+
+    def update_link_to(self, kind, version):
+        """Change the target of the link of the specified item to point
+        to the specified version. (Note that this method doesn't verify
+        that the specified version exists.)"""
+        if kind not in ALL_FOUR:
+            raise ValueError("unknown kind of item")
+        link = self.__getattribute__(kind)
+        filename = "{0}{1}.pem".format(kind, version)
+        # Relative rather than absolute target directory
+        target_directory = os.path.dirname(os.readlink(link))
+        # TODO: it could be safer to make the link first under a temporary
+        #       filename, then unlink the old link, then rename the new link
+        #       to the old link; this ensures that this process is able to
+        #       create symlinks.
+        # TODO: we might also want to check consistency of related links
+        #       for the other corresponding items
+        os.unlink(link)
+        os.symlink(os.path.join(target_directory, filename), link)
+
+    def update_all_links_to(self, version):
+        """Change the target of the cert, privkey, chain, and fullchain links
+        to point to the specified version."""
+        for kind in ALL_FOUR:
+            self.update_link_to(kind, version)
+
+    def notbefore(self, version=None):
+        """When is the beginning validity time of the specified version of the
+        cert in this lineage?  (If no version is specified, use the current
+        version.)"""
+        if version == None:
+            target = self.current_target("cert")
+        else:
+            target = self.version("cert", version)
+        pem = open(target).read()
+        x509 = OpenSSL.crypto.load_certificate(OpenSSL.crypto.FILETYPE_PEM,
+                                               pem)
+        i = x509.get_notBefore()
+        return pyrfc3339.parse(i[0:4] + "-" + i[4:6] + "-" + i[6:8] + "T" +
+                               i[8:10] + ":" + i[10:12] +":" +i[12:])
+
+    def notafter(self, version=None):
+        """When is the ending validity time of the specified version of the
+        cert in this lineage?  (If no version is specified, use the current
+        version.)"""
+        if version == None:
+            target = self.current_target("cert")
+        else:
+            target = self.version("cert", version)
+        pem = open(target).read()
+        x509 = OpenSSL.crypto.load_certificate(OpenSSL.crypto.FILETYPE_PEM,
+                                               pem)
+        i = x509.get_notAfter()
```

Comment:
Please deduplicate. `notafter` is exactly the same as `notbefore` apart from line above. `def notafter(self, version): return _notafterbefore(lambda x509: x509.get_notAfter)`?